internal/user/repository/view/model/user.go

```diff
@@ -298,6 +366,90 @@ func (u *UserView) setPasswordData(event *models.Event) error {
 	return nil
 }
 
+func (u *UserView) addPasswordLessToken(event *models.Event) error {
+	token, err := webAuthNViewFromEvent(event)
+	if err != nil {
+		return err
+	}
+	for _, t := range u.PasswordLessTokens {
+		if t.State == int32(model.MfaStateNotReady) {
+			t = token
+			return nil
+		}
+	}
+	u.U2FTokens = append(u.U2FTokens, token)
+	return nil
+}
+
+func (u *UserView) updatePasswordLessToken(event *models.Event) error {
+	token, err := webAuthNViewFromEvent(event)
+	if err != nil {
+		return err
+	}
+	for i, t := range u.PasswordLessTokens {
+		if t.ID == token.ID {
+			u.PasswordLessTokens[i].Name = token.Name
+			u.PasswordLessTokens[i].State = int32(model.MfaStateReady)
+			return nil
+		}
+	}
+	return nil
+}
+
+func (u *UserView) addU2FToken(event *models.Event) error {
+	token, err := webAuthNViewFromEvent(event)
+	if err != nil {
+		return err
+	}
+	for _, t := range u.U2FTokens {
+		if t.State == int32(model.MfaStateNotReady) {
+			t = token
+			return nil
+		}
+	}
+	u.U2FTokens = append(u.U2FTokens, token)
+	return nil
+}
+
+func (u *UserView) updateU2FToken(event *models.Event) error {
+	token, err := webAuthNViewFromEvent(event)
+	if err != nil {
+		return err
+	}
+	for i, t := range u.U2FTokens {
+		if t.ID == token.ID {
+			u.U2FTokens[i].Name = token.Name
+			u.U2FTokens[i].State = int32(model.MfaStateReady)
+			return nil
+		}
+	}
+	return nil
+}
+
+func (u *UserView) removeU2FToken(event *models.Event) error {
+	token, err := webAuthNViewFromEvent(event)
+	if err != nil {
+		return err
+	}
+	for i := len(u.U2FTokens) - 1; i >= 0; i-- {
+		if u.U2FTokens[i].ID == token.ID {
+			u.U2FTokens[i] = u.U2FTokens[len(u.U2FTokens)-1]
+			u.U2FTokens[len(u.U2FTokens)-1] = nil
+			u.U2FTokens = u.U2FTokens[:len(u.U2FTokens)-1]
+		}
+	}
+	return nil
+}
+
+func webAuthNViewFromEvent(event *models.Event) (*WebAuthNView, error) {
+	token := new(WebAuthNView)
+	err := json.Unmarshal(event.Data, token)
+	if err != nil {
+		return nil, caos_errs.ThrowInternal(err, "MODEL-FSaq1", "could not unmarshal data")
```

Comment:
Errors.Internal?

Reply:
what do you mean by `Errors.Internal`? this error handling is as any other unmarshalling in this or other view models

Reply:
i wasnt sure if we should return an error message Error.Internal instead of could not unmarshal, so we have translated errors

Reply:
this error happens in spooler... so there's no translation